client/asset/dcr/spv.go

```diff
@@ -0,0 +1,976 @@
+func (w *spvWallet) stop() {
+	w.log.Info("Unloading wallet")
+	if err := w.db.Close(); err != nil {
+		w.log.Info("Error closing database: %v", err)
+	}
+
+	w.log.Info("SPV wallet closed")
+}
+
+func (w *spvWallet) spvLoop(ctx context.Context, syncer *spv.Syncer) {
+	for {
+		err := syncer.Run(ctx)
+		if ctx.Err() != nil {
+			return
+		}
+		w.log.Errorf("SPV synchronization ended. trying again in 10 seconds: %v", err)
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Second * 10):
+		}
+	}
+}
+
+func (w *spvWallet) notesLoop(ctx context.Context, dcrw *wallet.Wallet) {
+	txNotes := dcrw.NtfnServer.TransactionNotifications()
+	defer txNotes.Done()
+	// removeTxNotes := dcrw.NtfnServer.RemovedTransactionNotifications()
+	// defer removeTxNotes.Done()
+	// acctNotes := dcrw.NtfnServer.AccountNotifications()
+	// defer acctNotes.Done()
+	// tipNotes := dcrw.NtfnServer.MainTipChangedNotifications()
+	// defer tipNotes.Done()
+	// confirmNotes := w.NtfnServer.ConfirmationNotifications(ctx)
+
+	for {
+		select {
+		case n := <-txNotes.C:
+			if len(n.AttachedBlocks) == 0 {
+				continue
+			}
+			lastBlock := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			h := lastBlock.Header.BlockHash()
+			select {
+			case w.tipChan <- &block{
+				hash:   &h,
+				height: int64(lastBlock.Header.Height),
+			}:
+			default:
+				w.log.Warnf("tip report channel was blocking")
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (w *spvWallet) tipFeed() <-chan *block {
+	return w.tipChan
+}
+
+// Connect starts the wallet and begins synchronization.
+func (w *spvWallet) Connect(ctx context.Context) error {
+	ctx, w.cancel = context.WithCancel(ctx)
+	err := w.startWallet(ctx)
+	if err != nil {
+		return err
+	}
+
+	w.wg.Add(1)
+	go func() {
+		defer w.wg.Done()
+		defer w.stop()
+
+		ticker := time.NewTicker(time.Minute * 20)
+
+		for {
+			select {
+			case <-ticker.C:
+				w.cleanBlockCache()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Disconnect shuts down the wallet and waits for monitored threads to exit.
+// Part of the Wallet interface.
+func (w *spvWallet) Disconnect() {
+	w.cancel()
+	w.wg.Wait()
+}
+
+// SpvMode is always true for spvWallet.
+// Part of the Wallet interface.
+func (w *spvWallet) SpvMode() bool {
+	return true
+}
+
+// NotifyOnTipChange is not used, in favor of the tipNotifier pattern from btc.
+func (w *spvWallet) NotifyOnTipChange(ctx context.Context, cb TipChangeCallback) bool {
+	return false
+}
+
+// AccountOwnsAddress checks if the provided address belongs to the specified
+// account.
+// Part of the Wallet interface.
+func (w *spvWallet) AccountOwnsAddress(ctx context.Context, addr stdaddr.Address, _ string) (bool, error) {
+	ka, err := w.KnownAddress(ctx, addr)
+	if err != nil {
+		return false, fmt.Errorf("KnownAddress error: %w", err)
+	}
+	if ka.AccountName() != w.acctName {
+		return false, nil
+	}
+	own, err := w.HaveAddress(ctx, addr)
```

Comment:
I was curious about these two `Wallet` methods:

```go
func (w *Wallet) HaveAddress(ctx context.Context, a stdaddr.Address) (bool, error) {
	const op errors.Op = "wallet.HaveAddress"
	err := walletdb.View(ctx, w.db, func(tx walletdb.ReadTx) error {
		addrmgrNs := tx.ReadBucket(waddrmgrNamespaceKey)
		_, err := w.manager.Address(addrmgrNs, a)
		return err
	})
	if err != nil {
		if errors.Is(err, errors.NotExist) {
			return false, nil
		}
		return false, errors.E(op, err)
	}
	return true, nil
}
```

```go
func (w *Wallet) KnownAddress(ctx context.Context, a stdaddr.Address) (KnownAddress, error) {
	const op errors.Op = "wallet.KnownAddress"

	var ma udb.ManagedAddress
	var acct uint32
	var acctName string
	err := walletdb.View(ctx, w.db, func(tx walletdb.ReadTx) error {
		addrmgrNs := tx.ReadBucket(waddrmgrNamespaceKey)
		var err error
		ma, err = w.manager.Address(addrmgrNs, a)
		if err != nil {
			return err
		}
		acct = ma.Account()
		acctName, err = w.manager.AccountName(addrmgrNs, acct)
		return err
	})
	if err != nil {
		return nil, errors.E(op, err)
	}

	var acctKind AccountKind
	switch {
	case acct < udb.ImportedAddrAccount:
		acctKind = AccountKindBIP0044
	case acct == udb.ImportedAddrAccount:
		acctKind = AccountKindImported
	default:
		acctKind = AccountKindImportedXpub
	}
	return wrapManagedAddress(ma, acctName, acctKind)
}
```

Some observations:
- `HaveAddress` seems to be a subset of `KnownAddress`
- `HaveAddress` doesn't actually do say if it is the "owner"
- we should check `ka.AccountKind() == wallet.AccountKindBIP0044` to ensure we actually have the private keys (the other kinds are `AccountKindImported` and `AccountKindImportedXpub`).  Maybe just exclude imported xpub?

Reply:
Important to accept `AccountKindImported` too?

Reply:
I suppose that would be needed for an address for an imported private key.
I commented on that other pr so hopefully it never means imported public key